pkg/store/tsdb.go

```diff
@@ -339,26 +339,42 @@ func (s *TSDBStore) LabelValues(ctx context.Context, r *storepb.LabelValuesReque
 	if err != nil {
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-
 	if !match {
 		return &storepb.LabelValuesResponse{}, nil
 	}
 
-	if v := s.getExtLset().Get(r.Label); v != "" {
-		for _, m := range matchers {
-			if !m.Matches(v) {
-				return &storepb.LabelValuesResponse{}, nil
-			}
-		}
-		return &storepb.LabelValuesResponse{Values: []string{v}}, nil
-	}
-
 	q, err := s.db.ChunkQuerier(r.Start, r.End)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	defer runutil.CloseWithLogOnErr(s.logger, q, "close tsdb querier label values")
 
+	// If we request label values for an external label while selecting an additional matcher for other label values
+	if val := s.getExtLset().Get(r.Label); val != "" {
+		if len(matchers) == 0 {
+			return &storepb.LabelValuesResponse{Values: []string{val}}, nil
+		}
+
+		hints := &storage.SelectHints{
+			Start: r.Start,
+			End:   r.End,
+			Func:  "series",
+		}
+		set := q.Select(ctx, false, hints, matchers...)
+
+		found := false
+		for set.Next() {
```

Comment:
Why do you think that you need to exhaust the reader here? :thinking: I don't think it's necessary.

Reply:
Converted it to the form you suggested earlier!